### PR TITLE
net: introduce IPv6 netstats

### PR DESCRIPTION
**Pull request by OlegHahm**
~~Based on #4801.~~

### CODE REVIEW

File: sys/include/net/gnrc/ipv6/netif.h [hunk 1]
```diff
@@ -587,6 +591,18 @@ static inline bool gnrc_ipv6_netif_addr_is_non_unicast(const ipv6_addr_t *addr)
...
+ * @brief   Get sent and received statistics about IPv6 traffic on this interface.
+ *
+ * @note    This function is only available if compiled with module `netstats_ipv6`.
+ *
+ * @param[in] pid   The PID to the interface.
+ *
+ * @return  A @ref netstats_t pointer to the statistics.
+ * @return  NULL if no statistics are available.
+ */
+netstats_t *gnrc_ipv6_netif_get_stats(kernel_pid_t pid);
```
Comment (cgundogan) [inline sys/include/net/gnrc/ipv6/netif.h]: Wouldn't it be more suitable to put this into some unit file of the netstats module itself? Then you can also drop the `@note` section and it is immediately apparent, that you need said module compiled into the binary.

File: sys/include/net/netstats.h [hunk 2]
```diff
@@ -18,13 +18,22 @@
...
 #define NETSTATS_H
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+typedef enum {
+    NETSTATS_LAYER2     = 0x01,
+    NETSTATS_IPV6       = 0x02,
+    NETSTATS_ALL        = 0xFF
```
Comment (cgundogan) [inline sys/include/net/netstats.h]: Not sure about the compiler optimization, but what about choosing `0` for `NETSTATS_ALL` to have contiguous numbers for the `switch case` below? This might increase the chances of using a jump table via function pointers by the copmiler, instead of a `if-else-if-` chain..

File: sys/include/net/netstats.h [hunk 3]
```diff
@@ -41,6 +50,16 @@ typedef struct {
     uint32_t rx_bytes;          /**< received bytes */
 } netstats_t;
 
+/**
+ * @brief   Get a string description for @ref netstats_module_t, for debugging
```
Comment (cgundogan) [inline sys/include/net/netstats.h]: Can you fix the indentation of the text after the doxygen command for this and the following two lines?

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 4]
```diff
@@ -386,6 +391,9 @@ static void _send_unicast(kernel_pid_t iface, uint8_t *dst_l2addr,
 
     DEBUG("ipv6: send unicast over interface %" PRIkernel_pid "\n", iface);
     /* and send to interface */
+#ifdef MODULE_NETSTATS_IPV6
+    gnrc_ipv6_netif_get_stats(iface)->tx_unicast_count++;
```
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: What about keeping the netstats struct cached in a (static) global (to this unit file only) variable, assigned only once in `gnrc_ipv6_init`? You can then reuse that at several places in this file, instead of calling `gnrc_ipv6_netif_get_stats` each time.

Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: I see, the `iface` parameter breaks this idea. But I still think this should be somehow cached .. most of the time we only have one interface.

File: sys/shell/commands/sc_netif.c [hunk 5]
```diff
@@ -73,30 +73,43 @@ static bool _is_iface(kernel_pid_t dev)
     return false;
 }
 
-#ifdef MODULE_NETSTATS_L2
-static int _netif_stats(kernel_pid_t dev, bool reset)
+#if defined(MODULE_NETSTATS_L2) || defined(MODULE_NETSTATS_IPV6)
```
Comment (cgundogan) [inline sys/shell/commands/sc_netif.c]: Just an idea: you could create a pseudo-module `MODULE_NETSTATS_PRINT` and let `MODULE_NETSTATS_L2`, `MODULE_NETSTATS_IPV6`, and hypothetically also future ones `MODULE_NETSTATS_UDP`, .. `MODULE_NETSTATS_X` depend on that. Then you can only check whether `MODULE_NETSTATS_PRINT` is defined instead of having a long and unreadable guard condition here.

File: sys/shell/commands/sc_netif.c [hunk 6]
```diff
@@ -989,11 +1006,36 @@ int _netif_config(int argc, char **argv)
             }
 #ifdef MODULE_NETSTATS_L2
```
Comment (cgundogan) [inline sys/shell/commands/sc_netif.c]: + `MODULE_NETSTATS_IPV6` (or `MODULE_NETSTATS_PRINT` if you wanna follow my advice from above)

File: sys/include/net/gnrc/ipv6/netif.h (hunk 1 above)
Comment (OlegHahm) [inline sys/include/net/gnrc/ipv6/netif.h]: IMO that would be conceptionally unclean. Actually, I think the better idea would be drop the netstats module completely and move its "functionality" into the shell.

File: sys/include/net/netstats.h (hunk 2 above)
Comment (OlegHahm) [inline sys/include/net/netstats.h]: Hm, IIRC the reason for choosing `0xFF` here was to enable checks like `if (module & NETSTATS_IPV6)` to return true for `NETSTATS_ALL`.

File: sys/include/net/netstats.h (hunk 3 above)
Comment (OlegHahm) [inline sys/include/net/netstats.h]: What do you mean by "fix"?

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 4 above)
Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: And in this case the call should just de-reference a pointer.

File: sys/shell/commands/sc_netif.c (hunk 5 above)
Comment (OlegHahm) [inline sys/shell/commands/sc_netif.c]: Checking for `NETSTATS` (for the current pseudomodule) should also do the trick. Yes, you're right.

File: examples/gnrc_networking/Makefile [hunk 7]
```diff
@@ -29,6 +29,8 @@ USEMODULE += gnrc_icmpv6_echo
 USEMODULE += shell
 USEMODULE += shell_commands
 USEMODULE += ps
+USEMODULE += netstats_l2
+USEMODULE += netstats_ipv6
```
Comment (OlegHahm) [inline examples/gnrc_networking/Makefile]: Not so sure if we really want this - should go into a separate PR.

File: sys/include/net/gnrc/ipv6/netif.h (hunk 1 above)
Comment (OlegHahm) [inline sys/include/net/gnrc/ipv6/netif.h]: What do you think about the new solution?

Comment (cgundogan) [inline sys/include/net/gnrc/ipv6/netif.h]: >...and move its "functionality" into the shell.

It depends: Can I still use netstats then if I do not compile the shell, but send the stats over e.g. UDP to a server?

File: sys/include/net/netstats.h (hunk 3 above)
Comment (cgundogan) [inline sys/include/net/netstats.h]: The description part of all three lines start at different indentations. I personally would like it more if you let them start at the same level.

File: sys/include/net/gnrc/ipv6/netif.h (hunk 1 above)
Comment (OlegHahm) [inline sys/include/net/gnrc/ipv6/netif.h]: Sure. The only thing in the `netstats` module was the `_to_string()` function that moved to the shell.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 4 above)
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: assuming the call gets inlined.

File: sys/include/net/netstats.h (hunk 3 above)
Comment (OlegHahm) [inline sys/include/net/netstats.h]: Of course, I can do this, but usually we don't do this for other headers.

File: examples/gnrc_networking/Makefile (hunk 7 above)
Comment (cgundogan) [inline examples/gnrc_networking/Makefile]: >Not so sure if we really want this - should go into a separate PR.

can you elaborate? I lost the context (:

File: sys/include/net/netstats.h (hunk 3 above)
Comment (OlegHahm) [inline sys/include/net/netstats.h]: Hm, well, actually I cannot do it - only for param and return, because otherwise brief would exceed the 80 character limit and would require a linebreak (which is not simple with doxygen).

Comment (OlegHahm) [inline sys/include/net/netstats.h]: addressed (as good as I can)

File: sys/include/net/netstats.h (hunk 2 above)
Comment (cgundogan) [inline sys/include/net/netstats.h]: okay

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 4 above)
Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Yes, we have to trust the compiler here a little bit. I'm not objecting "caching" this value, I just don't know how without sacrificing RAM.

File: examples/gnrc_networking/Makefile (hunk 7 above)
Comment (OlegHahm) [inline examples/gnrc_networking/Makefile]: I meant if we really want to have netstats per default in the gnrc_networking example.

File: sys/include/net/netstats.h (hunk 3 above)
Comment (cgundogan) [inline sys/include/net/netstats.h]: >would exceed the 80 character limit and would require a linebreak (which is not simple with doxygen)

since when? AFAIK, you can just put in a line break and move the text for the second line again to the correct indentation level. No need for magic.

Comment (OlegHahm) [inline sys/include/net/netstats.h]: Apparently I remembered that one wrong (or got confused by vim's syntax highlighting), but the discussion is obsolete anyway - the function is gone. :)

File: examples/gnrc_networking/Makefile (hunk 7 above)
Comment (cgundogan) [inline examples/gnrc_networking/Makefile]: I see. I wouldn't object, because `gnrc_networking` is IMO for showcasing our features, while `gnrc_minimal` is for showcasing our minamility.

File: sys/include/net/netstats.h (hunk 2 above)
Comment (authmillenon) [inline sys/include/net/netstats.h]: > Hm, IIRC the reason for choosing 0xFF here was to enable checks like if (module & NETSTATS_IPV6) to return true for NETSTATS_ALL.

Then you should not use an enum IMHO, since `&` with enums isn't working IIRC.

Comment (OlegHahm) [inline sys/include/net/netstats.h]: It seems to work and I cannot think of any reason why it should not.

Comment (authmillenon) [inline sys/include/net/netstats.h]: Okay, maybe this was more of a signed/unsigned issue.... Can't remember, just that we had (casting related) issues with flags defined in enums in the past, that yielded compiler warnings.

File: examples/gnrc_networking/Makefile (hunk 7 above)
Comment (cgundogan) [inline examples/gnrc_networking/Makefile]: What about compiling in `netstats_ipv6` automatically if `netstats` and `ipv6` are compiled?

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 8]
```diff
@@ -340,6 +340,11 @@ static void _send_to_iface(kernel_pid_t iface, gnrc_pktsnip_t *pkt)
         gnrc_pktbuf_release(pkt);
         return;
     }
+#ifdef MODULE_NETSTATS_IPV6
+    if_entry->stats.tx_success++;
+    if_entry->stats.tx_bytes += gnrc_pkt_len(pkt->next);
```
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: not sure if it's available here, but why not use the length field of the `ipv6` header? Isn't that more accurate than `gnrc_pkt_len(pkt->next)`?

Comment (authmillenon) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: IPv6 header length + ipv6->len; that would be. `pkt->next` may contain the netif header that isn't really data that is send over the interface.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c [hunk 9]
```diff
@@ -772,6 +783,13 @@ static void _receive(gnrc_pktsnip_t *pkt)
     assert(netif);
     iface = ((gnrc_netif_hdr_t *)netif->data)->if_pid;
 
+#ifdef MODULE_NETSTATS_IPV6
+    assert(iface);
+    netstats_t *stats = gnrc_ipv6_netif_get_stats(iface);
+    stats->rx_count++;
+    stats->rx_bytes += (gnrc_pkt_len(pkt) - netif->size);
```
Comment (authmillenon) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Why do you add the netif->size here?

File: examples/gnrc_networking/Makefile (hunk 7 above)
Comment (OlegHahm) [inline examples/gnrc_networking/Makefile]: Hm, what if I'm only interested in L2 stats?

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 9 above)
Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: I'm subtracting it.

**Comment by OlegHahm** [comment]
Btw I don't know how to handle traffic from unknown interfaces.

File: examples/gnrc_networking/Makefile (hunk 7 above)
Comment (cgundogan) [inline examples/gnrc_networking/Makefile]: `DISABLE_MODULE=netstats_ipv6`. But it was just food for thought. I am fine with both variants.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 8 above)
Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: I agree that `IPv6 header length + ipv6->len` sounds like a good idea, but in which case can `pkt->next` contain a netif header and which type would `pkt` be in this case?

**Comment by cgundogan** [comment]
>Btw I don't know how to handle traffic from unknown interfaces.

IMO, if there is no interface, then just skip the statistics for those packets.

**Comment by cgundogan** [comment]
>IMO, if there is no interface, then just skip the statistics for those packets.

On the other hand, you could also aggregate the statistics for `KERNEL_PID_UNDEF` in a seperate struct

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 8 above)
Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Hm, looking at the code again, I wonder if the current solution is not more reasonable: we add the amount of data that is actually passed to the lower layer here.

**Comment by OlegHahm** [comment]
If we had a separate dedicated descriptors for pseudo interfaces (like loopback), netstats could help to find bugs like https://github.com/RIOT-OS/RIOT/issues/5257.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 8 above)
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: But this amount of data can also include padding, if I am not mistaken? AFAIK, the pkts in the pktbuf need to be size aligned(?)

Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Hm, who does actually remove the padding before sending?

Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Hm, no idea. In [gnrc_pktbuf_static.c:423](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c#L423) the size for a `pktnsip` gets aligned, but I have no idea where the magic happens that `gnrc_pkt_len` reports the correct size. Maybe in [gnrc_pktbuf_static.c:449](https://github.com/RIOT-OS/RIOT/blob/master/sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c#L449)?

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 9 above)
Comment (authmillenon) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: That comes from reviewing before coffee ^^.

Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: Anyway this has to be changed if #5179 is reverted.

File: sys/include/net/netstats.h (hunk 2 above)
Comment (cgundogan) [inline sys/include/net/netstats.h]: Actually, I don't know whether @authmillenon's statement is correct or not,
but it's also written in our coding conventions:
[Thou shalt not use enums as flags](https://github.com/RIOT-OS/RIOT/wiki/Best-Practice-for-RIOT-Programming#donts)

Comment (OlegHahm) [inline sys/include/net/netstats.h]: Who am I to question the BCPs? Addressed.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 9 above)
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: >Anyway this has to be changed if #5179 is reverted.

So, it needs to be changed now?

File: sys/shell/commands/sc_netif.c [hunk 10]
```diff
@@ -987,13 +1018,38 @@ int _netif_config(int argc, char **argv)
...
+                    module = NETSTATS_ALL;
+                }
+                else if (strcmp(argv[3], "l2") == 0) {
+                    module = NETSTATS_LAYER2;
+                }
+                else if (strcmp(argv[3], "ipv6") == 0) {
+                    module = NETSTATS_IPV6;
+                }
+                else {
+                    printf("Module %s doesn't exist or do not provide statistics.\n", argv[3]);
```
Comment (cgundogan) [inline sys/shell/commands/sc_netif.c]: s/do/does/

**Comment by cgundogan** [comment]
I guess for the first iteration everything seems fine. ACK once you addressed my comments. Please squash.

File: sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c (hunk 9 above)
Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: `sizeof(netif_hdr_t)` is maybe more apparent than `netif->size`

Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: But hasn't this a different semantic since the L2 addresses are not part of the struct?

Comment (cgundogan) [inline sys/net/gnrc/network_layer/ipv6/gnrc_ipv6.c]: ah you are right. nevermind then.

**Comment by OlegHahm** [comment]
Sorry for the delay. Addressed remaining issues and rebased to master.

**Comment by authmillenon** [comment]
Can you squash then, please?

**Comment by OlegHahm** [comment]
Squashed into a single commit.

**Comment by authmillenon** [comment]
There is an error on some boards in the `netstats` app.

**Comment by OlegHahm** [comment]
The last commit should fix it.

**Comment by OlegHahm** [comment]
okay to squash?

**Comment by miri64** [comment]
Yes. ACK, as soon as Murdock is happy.
